### DISCUSSION TITLE
enable ssl protocol to the ice endpoints

### DISCUSSION THREAD

**Discussion by karthikvadapalli** [zeroc-ice/ice#2471] [Q&A]
Hi our application is using Ice3.3.1 and 3.4.2(due to the linux versions) with the tcp protocols, below is our configuration

                                <ice.bin>/icestation/Ice-3.4.2/bin</ice.bin>
				<icestorm.service-entry-point>IceStormService,34:createIceStorm</icestorm.service-entry-point>
				<icebox.servicemanager.endpoints>tcp -p 10008</icebox.servicemanager.endpoints>
				<icestorm.topicmanager.endpoints>default -h hostname -p 10010</icestorm.topicmanager.endpoints>
				<icestorm.publish.endpoints>tcp -p 10011:udp -p 10011</icestorm.publish.endpoints>
				<icestorm.assembly.id>assemblyID</icestorm.assembly.id>

now we want to make it work with SSL protocol, can this be possible, if yes how we need to implement the solution?

**Comment by pepone** [comment]
Ice 3.4 is quite old, you need to upgrade to 3.7.

Using SSL is a matter of adding an SSL endpoint, and configuring IceSSL plug-in:

See 3.4 [docs](https://doc.zeroc.com/display/Ice34/IceSSL) and [demo applications](https://github.com/zeroc-ice/ice-demos/tree/3.4)

**Reply by karthikvadapalli** [reply]
thanks Pepone, do you have a working example for Java?

**Reply by karthikvadapalli** [reply]
we are creating our icestorm instance like below

mkdir ${icestorm.db.path}
sc create Icestorm binPath= "${ice.bin}\icebox.exe --Ice.Config=${icestorm.home}\config\config.icebox --service icestorm2" obj= ${icegridnode.service.user} password= ${icegridnode.service.password}
sc start Icestorm

**Reply by pepone** [reply]
The `demoj` directory in the linked repository contains the Java examples for Ice 3.4.2.

**Reply by karthikvadapalli** [reply]
sorry for too many questions, i am new to zeroc development.

i don't see icessl implementation in the demoj folder

**Reply by bernardnormier** [reply]
IceSSL is transport plugin.  In order to use SSL with Ice, you typically just need to:
 - configure this plugin (see for example https://github.com/zeroc-ice/ice-demos/blob/c5091b705bd74e2925b249f92e8fc7b3a7769336/demoj/Ice/hello/config.client#L41)
 - use a proxy / object adapter with a `ssl` endpoint

That's what the hello demo does but unfortunately not in a straightforward manner.

Please note the ssl endpoint and Ice.Plugin.IceSSL configuration in the server:
https://github.com/zeroc-ice/ice-demos/blob/c5091b705bd74e2925b249f92e8fc7b3a7769336/demoj/Ice/hello/config.server#L6

On the client side, the proxy with an ssl endpoint comes from:
https://github.com/zeroc-ice/ice-demos/blob/c5091b705bd74e2925b249f92e8fc7b3a7769336/demoj/Ice/hello/config.client#L5

**Reply by karthikvadapalli** [reply]
added the below configuration into config.service 
#
#Ice SSL
#
Ice.Plugin.IceSSL=IceSSL.PluginFactory
IceSSL.DefaultDir=${icestrom.home}
IceSSL.Keystore=cacerts.jks
IceSSL.Password=password
IceSSL.Truststore=cacerts.jks

while starting the icestorm i am getting below error, can you please help

!! 07/22/24 18:10:00.941 /icestation/Ice-3.4.2/bin/icebox: error: ServiceManager: exception while starting service IceStorm:
   PluginManagerI.cpp:357: Ice::PluginInitializationException:
   plug-in initialization failed: unable to load entry point `IceSSL.PluginFactory': invalid entry point format `IceSSL.PluginFactory'

**Reply by pepone** [reply]
IceStorm is a C++ service, you need to configure the C++ IceSSL plug-in. See for example the C++ hello demo configuration:

https://github.com/zeroc-ice/ice-demos/blob/c5091b705bd74e2925b249f92e8fc7b3a7769336/demo/Ice/hello/config.server#L42-L46

**Reply by karthikvadapalli** [reply]
can you able to help with this error

ERROR | 2024-07-23T11:09:31,205 | [main] IceGridNode -- Exception thrown while starting IceGridNode
Ice.EndpointParseException: null
        at Ice.ObjectAdapterI.parseEndpoints(ObjectAdapterI.java:1120) ~[ice-3.3.1.jar:?]
        at Ice.ObjectAdapterI.<init>(ObjectAdapterI.java:901) ~[ice-3.3.1.jar:?]
        at IceInternal.ObjectAdapterFactory.createObjectAdapter(ObjectAdapterFactory.java:183) ~[ice-3.3.1.jar:?]
        at Ice.CommunicatorI.createObjectAdapterWithEndpoints(CommunicatorI.java:77) ~[ice-3.3.1.jar:?]
        at com.db.icestation.server.IceGridNode.run(IceGridNode.java:46) [server-9.6-SNAPSHOT.jar:?]
        at Ice.Application.main(Application.java:116) [ice-3.3.1.jar:?]
        at com.db.icestation.server.IceGridNode.main(IceGridNode.java:38) [server-9.6-SNAPSHOT.jar:?]

earlier the endpoint uri is default -p 4071 now i changed it to ssl -p 4071 which is throwing the error

**Reply by pepone** [reply]
Did you also configure the IceSSL plugin in the Java application raising this error?

**Reply by karthikvadapalli** [reply]
it is failing at the below location
ObjectAdapter adapter =
                    communicator().createObjectAdapterWithEndpoints("IcestationAdapter", "ssl -p 4071");

**Reply by pepone** [reply]
This means the IceSSL plugin is not configured in the communicator.

**Reply by karthikvadapalli** [reply]
we have config.icebox and config.service files are present which i added the icsSSL plugin
do you have an example where the iceSSL added to the communicator?

**Reply by karthikvadapalli** [reply]
> IceStorm is a C++ service, you need to configure the C++ IceSSL plug-in. See for example the C++ hello demo configuration:
> 
> https://github.com/zeroc-ice/ice-demos/blob/c5091b705bd74e2925b249f92e8fc7b3a7769336/demo/Ice/hello/config.server#L42-L46

can you help on how to create cert.pem c_rsa1024_pub.pem and c_rsa1024_priv.pem files on my own?

**Reply by pepone** [reply]
There are multiple tutorials on the web, google for "how to create custom ssl certificate" or "How to create self signed certificates"

**Reply by karthikvadapalli** [reply]
we are connecting from our java application to icestorm added the selfsigned certs on the icestorm side but java side it is only expecting jks file which jks file we need to add to the ICESSL configuration?

**Reply by pepone** [reply]
You need to at least add a JKS Key store with the CA certificate, for the Java application to be able to verify the IceStorm certificate.

**Reply by karthikvadapalli** [reply]
i am trying the client.jks file from the examples and getting the below error
   error # = 336151571
   message = error:14094413:SSL routines:SSL3_READ_BYTES:sslv3 alert unsupported certificate
   location = s3_pkt.c, 1275
   data = SSL alert number 43

**Reply by pepone** [reply]
Which certificate?

**Reply by karthikvadapalli** [reply]
https://github.com/zeroc-ice/ice-demos/blob/c5091b705bd74e2925b249f92e8fc7b3a7769336/certs/client.jks

**Reply by pepone** [reply]
The certificate expired, you need to create new certificates.

For testing you can use the certificates from https://github.com/zeroc-ice/ice/tree/main/certs

**Reply by karthikvadapalli** [reply]
Hi 
we are creating an adapter on client side with in the run method like below.
communicator().createObjectAdapterWithEndpoints("adapter","ssl -p 10010)

our main class is below:
Map<String, String> props = (Map<String, String>)implementationFactory.getImplFor("IceCommunicatorProperties");
        InitializationData initData = new InitializationData();
        initData.properties = Util.createProperties();
        for (Map.Entry<String, String> e : props.entrySet()) {
            initData.properties.setProperty(e.getKey(), e.getValue());
        }
        int status = new IceGridNode().main("IceStation", args, initData);
        LOG.info("IceGridNode Status:" + status);
        // start Ice communicator
        System.exit(status);

below is the IceCommunicatorProperties bean definition:

<bean id="IceCommunicatorProperties" class="org.springframework.beans.factory.config.MapFactoryBean">
		<property name="sourceMap">
			<map>
				<entry key="Ice.Default.Package" value="com.db.icestation.ice"/>
				<entry key="Ice.MessageSizeMax" value="256000"/>
				<entry key="Ice.Plugin.IceSSL" value="IceSSL.PluginFactory"/>
				<entry key="IceSSL.DefaultDir" value="/ssl/"/>
				<entry key="IceSSL.Keystore" value="server.jks"/>
				<entry key="IceSSL.Truststore" value="ca.jks"/>
				<entry key="IceSSL.Password" value="changeit"/>
			</map>
		</property>
	</bean>

while connecting from client with the same above configuration we are getting no cipher suites in common, any help would be appreciated

**Reply by pepone** [reply]
"no cipher suites in common" means that the client and the server cannot agree on SSL cipher suite that is accepted by both, and the TLS handshakes cannot continue.

The available cipher suites depend on the SSL platforms, for Java it depends on the Java version you are using, and for the C++ side on Linux it depends on the OpenSSL side.

There is also IceSSL.Cipher property to filter-out ciphers, but I would left this out, just ensure you are not setting it to a restricted value that is not available by the peer.

You should really try to upgrade to a more recent Ice versions, one thing that could be happening is that the OS is configured to disallow old SSL protocols, and this old Ice versions is still trying to use them.

**Reply by karthikvadapalli** [reply]
how can i check what ciphers ice ssl is using and java is using

the java version we are using is 1.8.0_401 and OS is RHEL 6.10

i setup the flag IceSSL.Trace.Security to 2 am getting the below enabled ciphers:
[ 02/10/24 15:53:59:958 Security: enabling SSL ciphersuites:
    TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384
    TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384
    TLS_RSA_WITH_AES_256_CBC_SHA256
    TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA384
    TLS_ECDH_RSA_WITH_AES_256_CBC_SHA384
    TLS_DHE_RSA_WITH_AES_256_CBC_SHA256
    TLS_DHE_DSS_WITH_AES_256_CBC_SHA256
    TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA
    TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
    TLS_RSA_WITH_AES_256_CBC_SHA
    TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA
    TLS_ECDH_RSA_WITH_AES_256_CBC_SHA
    TLS_DHE_RSA_WITH_AES_256_CBC_SHA
    TLS_DHE_DSS_WITH_AES_256_CBC_SHA
    TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256
    TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
    TLS_RSA_WITH_AES_128_CBC_SHA256
    TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA256
    TLS_ECDH_RSA_WITH_AES_128_CBC_SHA256
    TLS_DHE_RSA_WITH_AES_128_CBC_SHA256
    TLS_DHE_DSS_WITH_AES_128_CBC_SHA256
    TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA
    TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
    TLS_RSA_WITH_AES_128_CBC_SHA
    TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA
    TLS_ECDH_RSA_WITH_AES_128_CBC_SHA
    TLS_DHE_RSA_WITH_AES_128_CBC_SHA
    TLS_DHE_DSS_WITH_AES_128_CBC_SHA
    TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
    TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
    TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
    TLS_RSA_WITH_AES_256_GCM_SHA384
    TLS_ECDH_ECDSA_WITH_AES_256_GCM_SHA384
    TLS_ECDH_RSA_WITH_AES_256_GCM_SHA384
    TLS_DHE_RSA_WITH_AES_256_GCM_SHA384
    TLS_DHE_DSS_WITH_AES_256_GCM_SHA384
    TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
    TLS_RSA_WITH_AES_128_GCM_SHA256
    TLS_ECDH_ECDSA_WITH_AES_128_GCM_SHA256
    TLS_ECDH_RSA_WITH_AES_128_GCM_SHA256
    TLS_DHE_RSA_WITH_AES_128_GCM_SHA256
    TLS_DHE_DSS_WITH_AES_128_GCM_SHA256
    TLS_ECDHE_ECDSA_WITH_3DES_EDE_CBC_SHA
    TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA
    SSL_RSA_WITH_3DES_EDE_CBC_SHA
    TLS_ECDH_ECDSA_WITH_3DES_EDE_CBC_SHA
    TLS_ECDH_RSA_WITH_3DES_EDE_CBC_SHA
    SSL_DHE_RSA_WITH_3DES_EDE_CBC_SHA
    SSL_DHE_DSS_WITH_3DES_EDE_CBC_SHA
    TLS_EMPTY_RENEGOTIATION_INFO_SCSV ]

**Reply by karthikvadapalli** [reply]
i can see the ciphers on java are matching with most of the above

**Reply by karthikvadapalli** [reply]
application is struck at the below point, do you have any suggestions to implement?
communicator.stringToProxy("Hello:ssl -h hostname -p 10010");

**Comment by pepone** [comment]
If you are using the Ice.Application class as show in the stack above, the configuration file used to create the communicator can be passed as an argument to Application.main

See https://github.com/zeroc-ice/ice-demos/blob/c5091b705bd74e2925b249f92e8fc7b3a7769336/demoj/Ice/hello/Server.java#L34

**Reply by karthikvadapalli** [reply]
i am sending the below parameters while initiating the communicator but it is not getting proxy to connect, is the below attributes sufficient or need more?
```
<bean id="IceCommunicatorProperties" class="org.springframework.beans.factory.config.MapFactoryBean">
		<property name="sourceMap">
			<map>
				<entry key="Ice.Default.Package" value="com.db.icestation.ice"/>
				<entry key="Ice.MessageSizeMax" value="256000"/>
				<entry key="Ice.Plugin.IceSSL" value="IceSSL.PluginFactory"/>
				<entry key="IceSSL.DefaultDir" value="/icestation/UAT"/>
				<entry key="IceSSL.Keystore" value="cacerts.jks"/>
				<entry key="IceSSL.password" value="${keystore.password}"/>
			</map>
		</property>
	</bean>
```

**Reply by pepone** [reply]
How are you passing this properties to Ice.Application? The XML above doesn't seem like something Ice.Application would read out-of the box.

**Reply by karthikvadapalli** [reply]
i am setting these properties to InitializationData and using to initialize communicator

**Reply by pepone** [reply]
Can you show us the code, or better create a self contained example that shows your problem?

**Reply by karthikvadapalli** [reply]
how to show you the code?
below is the log which shows the proxy is connected but it is unable to activate the listener

INFO 2024-07-30 09:18:36,781 [main] IceAdminGuiServiceImpl: Set IceStationServer.Primary=Icestation:ssl -h hostname -p 4071:tcp -h hostname -p 4091
INFO 2024-07-30 09:18:36,781 [main] IceAdminGuiServiceImpl: Set IceStationServer.Secondary=Icestation:ssl -h hostname -p 4081:tcp -h hostname -p 4051
INFO 2024-07-30 09:18:36,828 [main] IceStationServerConnection: Created connection instance IceStationServer.Primary
INFO 2024-07-30 09:18:40,943 [main] IceStationServerConnection: Created connection instance IceStationServer.Secondary

**Reply by karthikvadapalli** [reply]
lookslike below is not working,
ObjectPrx p = communicator.propertyToProxy(IceStationServer.Primary);
            if (p == null) {
                return false;
            }
            // Disable the locator cache of this proxy: every time a connection is closed, a new
            // locate request is made to determine the endpoint for the new connection. Connections
            // are closed automatically when they time out, and by us when we detect a change in one
            // of the object adapters.
            // Note that proxies are immutable, that is why the below call returns a new proxy.
            p = p.ice_locatorCacheTimeout(0);
            // Cast the proxy to its known type.
            proxy = castProxy(p);

**Comment by pepone** [comment]
> i am setting these properties to InitializationData and using to initialize communicator

It would be simpler if you upload a zip file with the code that demostrate the problem, so we can take a look. otherwise is difficult to diagnose the issue.

**Reply by karthikvadapalli** [reply]
is it necessary to add tcp/udp ports along with ssl ports?
if yes how do we know it is encrypted and using ssl?

**Reply by pepone** [reply]
> is it necessary to add tcp/udp ports along with ssl ports?

No having a single ssl endpoint is fine, and better if you want to ensure there is only secure connections.

**Reply by karthikvadapalli** [reply]
i am getting below error when i am using the DemoIceStorm/TopicManager:ssl -h hostname -p 10020
Couldn't connect to IceStorm
Ice.ConnectionRefusedException: null
        at IceInternal.Network.doFinishConnect(Network.java:417) ~[ice-3.3.1.jar:?]
        at IceSSL.TransceiverI.initialize(TransceiverI.java:37) ~[ice-3.3.1.jar:?]
        at Ice.ConnectionI.initialize(ConnectionI.java:1536) ~[ice-3.3.1.jar:?]
        at Ice.ConnectionI.socketReady(ConnectionI.java:1116) ~[ice-3.3.1.jar:?]
        at Ice.ConnectionI$SocketReadyCallback.socketReady(ConnectionI.java:2299) ~[ice-3.3.1.jar:?]
        at IceInternal.SelectorThread.run(SelectorThread.java:203) ~[ice-3.3.1.jar:?]
        at IceInternal.SelectorThread$HelperThread.run(SelectorThread.java:273) ~[ice-3.3.1.jar:?]

**Reply by pepone** [reply]
Did you configure IceStorm to listen on `ssl -h hostname -p 10020` endpoint?

**Reply by karthikvadapalli** [reply]
-- 09/23/24 14:25:00.056 /opt/Ice-3.4.2/bin/icebox-IceStorm: Network: failed to establish ssl connection
   local address: ipaddress:10020
   remote address: :-1
   TransceiverI.cpp:234: Ice::ProtocolException:
   protocol exception:
   SSL error occurred for new incoming connection:
   remote address = ipaddress:33316
   no shared cipher

this is what i can see when i am starting the icestorm process

**Reply by pepone** [reply]
> no shared cipher

This means that the client and server cannot negotiate a common SSL cipher to use. There can be a number of reason why this happens, for example OpenSSL and Java version used in your setup. Setting `IceSSL.Trace.Security=1` might give you additional info about what ciphers are available in each application.

The best would be to upgrade your application to a recent Ice release, if you cannot and need commercial support for this older versions try contact info@zeroc.com for available support options.

Cheers,
Jose

**Reply by karthikvadapalli** [reply]
error # = 336109761
   message = error:1408A0C1:SSL routines:SSL3_GET_CLIENT_HELLO:no shared cipher
   location = s3_srvr.c, 1407

this is the only information i am getting after adding IceSSL.Trace.Security=1

**Reply by karthikvadapalli** [reply]
is there any restrictions of java versions or  Openssl we need to use to connect to icessl?
currently we are using java 1.8 and OpenSSL 1.0.1e-fips 11 Feb 2013

**Reply by karthikvadapalli** [reply]
any help with the below error
Couldn't connect to IceStorm
Ice.SecurityException: null
        at IceSSL.TransceiverI.handshakeNonBlocking(TransceiverI.java:454) ~[ice-3.3.1.jar:?]
        at IceSSL.TransceiverI.initialize(TransceiverI.java:43) ~[ice-3.3.1.jar:?]
        at Ice.ConnectionI.initialize(ConnectionI.java:1536) ~[ice-3.3.1.jar:?]
        at Ice.ConnectionI.socketReady(ConnectionI.java:1116) ~[ice-3.3.1.jar:?]
        at Ice.ConnectionI$SocketReadyCallback.socketReady(ConnectionI.java:2299) ~[ice-3.3.1.jar:?]
        at IceInternal.SelectorThread.run(SelectorThread.java:203) ~[ice-3.3.1.jar:?]
        at IceInternal.SelectorThread$HelperThread.run(SelectorThread.java:273) ~[ice-3.3.1.jar:?]
Caused by: javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
        at sun.security.ssl.Alert.createSSLException(Alert.java:131) ~[?:1.8.0_401]
        at sun.security.ssl.Alert.createSSLException(Alert.java:117) ~[?:1.8.0_401]
        at sun.security.ssl.TransportContext.fatal(TransportContext.java:342) ~[?:1.8.0_401]
        at sun.security.ssl.Alert$AlertConsumer.consume(Alert.java:293) ~[?:1.8.0_401]
        at sun.security.ssl.TransportContext.dispatch(TransportContext.java:185) ~[?:1.8.0_401]
        at sun.security.ssl.SSLTransport.decode(SSLTransport.java:156) ~[?:1.8.0_401]
        at sun.security.ssl.SSLEngineImpl.decode(SSLEngineImpl.java:588) ~[?:1.8.0_401]
        at sun.security.ssl.SSLEngineImpl.readRecord(SSLEngineImpl.java:544) ~[?:1.8.0_401]
        at sun.security.ssl.SSLEngineImpl.unwrap(SSLEngineImpl.java:411) ~[?:1.8.0_401]
        at sun.security.ssl.SSLEngineImpl.unwrap(SSLEngineImpl.java:390) ~[?:1.8.0_401]
        at javax.net.ssl.SSLEngine.unwrap(SSLEngine.java:629) ~[?:1.8.0_401]
        at IceSSL.TransceiverI.handshakeNonBlocking(TransceiverI.java:380) ~[ice-3.3.1.jar:?]
        ... 6 more

**Reply by karthikvadapalli** [reply]
can we use the Ice-3.4.2 in GCP cloud?
